### PR TITLE
Insights/user permissions

### PR DESCRIPTION
**Pull request by coury-clark**
Closes https://github.com/sourcegraph/sourcegraph/issues/24438

To summarize this PR, previously any backend insights were unrestricted based on user / org permissions, and the GraphQL API would return all insights to all users. In preparation for building an API for insights and dashboards, we obviously had to solve this problem.

1. Permission grant associations for insight views to user, org, and global
    1. Eventually this will get partitioned into more granular permission levels, but for now everyone is assumed to have read / write to match feature parity of frontend insights.
    2. Insights defined in a users private settings will get associated with that user ID
    3. Insights defined in an org will get associated with that org ID, and any user in that org can view them
2. The sync job from settings -> the database had to be updated to capture permissions updates as insights move between dashboards.
    1. The sync job will now just optimistically delete and replace any insight view that is defined in settings. This will still leave dangling insights that have been removed from the dashboard, but that's okay for the interm without an API.

New DDL looks like this:
CleanShot 2021-09-15 at 15 38 39@2x

### CODE REVIEW

File: enterprise/internal/insights/dbtesting/insights.go [hunk 1]
```diff
@@ -50,6 +52,7 @@ func TimescaleDB(t testing.TB) (db *sql.DB, cleanup func()) {
 
 	// Create database just for this test.
 	dbname := "insights_test_" + strings.ToLower(strings.ReplaceAll(t.Name(), "/", "_"))
+	log15.Info("dbname", "name", dbname)
```
Comment (coury-clark) [inline enterprise/internal/insights/dbtesting/insights.go:55]:
```suggestion
```

File: enterprise/internal/insights/discovery/discovery.go [hunk 2]
```diff
@@ -155,35 +155,30 @@ func (m *settingMigrator) migrate(ctx context.Context) error {
 		return err
 	}
 
-	var count, skipped, errors int
+	var count, skipped, errorCount int
```
Comment (coury-clark) [inline enterprise/internal/insights/discovery/discovery.go:158]: Just renaming to avoid package clash, unrelated to changes in this PR

File: enterprise/internal/insights/discovery/discovery.go [hunk 3]
```diff
@@ -155,35 +155,30 @@ func (m *settingMigrator) migrate(ctx context.Context) error {
...
 			continue
 		}
 
 		err = migrateSeries(ctx, insightStore, d)
 		if err != nil {
 			// we can't do anything about errors, so we will just skip it and log it
-			errors++
-			log15.Error("error while migrating insight", "error", err)
+			errorCount++
+			log15.Error("insights migraiton: error while migrating insight", "error", err)
```
Comment (coury-clark) [inline enterprise/internal/insights/discovery/discovery.go:177]:
```suggestion
			log15.Error("insights migration: error while migrating insight", "error", err)
```

File: enterprise/internal/insights/resolvers/insight_connection_resolver.go [hunk 4]
```diff
@@ -67,7 +72,25 @@ func (r *insightConnectionResolver) PageInfo(ctx context.Context) (*graphqlutil.
 
 func (r *insightConnectionResolver) compute(ctx context.Context) ([]types.Insight, int64, error) {
 	r.once.Do(func() {
-		mapped, err := r.insightMetadataStore.GetMapped(ctx, store.InsightQueryArgs{UniqueIDs: r.ids})
+		args := store.InsightQueryArgs{UniqueIDs: r.ids}
+		uid := actor.FromContext(ctx).UID
+		if uid != 0 {
+			// 🚨 SECURITY
+			// only filter user if the actor is not anonymous so that internal users of the API have super-admin like behavior
```
Comment (coury-clark) [inline enterprise/internal/insights/resolvers/insight_connection_resolver.go:79]: Reading this comment suddenly doesn't make any sense to me. I think the behavior is correct (anonymous users don't get to see anything) but I'll revisit this and check.

Comment (coury-clark) [reply enterprise/internal/insights/resolvers/insight_connection_resolver.go]: I added a test to assert the behavior is correct (it is) and changed the comment to make more sense.

File: enterprise/internal/insights/store/permissions.go [hunk 5]
```diff
@@ -61,3 +61,28 @@ func NewInsightPermissionStore(db dbutil.DB) *InsightPermStore {
...
+func UserGrant(userID int) InsightViewGrant {
+	return InsightViewGrant{UserID: &userID}
+}
+
+func OrgGrant(orgID int) InsightViewGrant {
+	return InsightViewGrant{OrgID: &orgID}
+}
+
+func GlobalGrant() InsightViewGrant {
+	b := true
```
Comment (efritz) [inline enterprise/internal/insights/store/permissions.go:86]: image

File: enterprise/internal/insights/store/insight_store.go [hunk 6]
```diff
@@ -47,11 +47,13 @@ func (s *InsightStore) Transact(ctx context.Context) (*InsightStore, error) {
...
 	UniqueIDs []string
 	UniqueID  string
+	UserID    []int
+	OrgID     []int
 }
 
 // Get returns all matching viewable insight series.
 func (s *InsightStore) Get(ctx context.Context, args InsightQueryArgs) ([]types.InsightViewSeries, error) {
-	preds := make([]*sqlf.Query, 0, 2)
+	preds := make([]*sqlf.Query, 0, 4)
```
Comment (efritz) [inline enterprise/internal/insights/store/insight_store.go:56]: I think you can only have 1, 2, or 3 elems in this slice (but is such a nitty comment).

File: enterprise/internal/insights/store/insight_store.go [hunk 7]
```diff
@@ -72,6 +75,28 @@ func (s *InsightStore) Get(ctx context.Context, args InsightQueryArgs) ([]types.
...
+	if len(args.UserID) > 0 {
+		elems := make([]*sqlf.Query, 0, len(args.UserID))
+		for _, id := range args.UserID {
+			elems = append(elems, sqlf.Sprintf("%s", id))
+		}
+		permsPreds = append(permsPreds, sqlf.Sprintf("ivg.user_id IN (%s)", sqlf.Join(elems, ",")))
+	}
+	permsPreds = append(permsPreds, sqlf.Sprintf("ivg.global is true"))
+
+	return sqlf.Sprintf("(%s)", sqlf.Join(permsPreds, " OR "))
```
Comment (efritz) [inline enterprise/internal/insights/store/insight_store.go:97]: FYI `sqlf.Join` adds the whitespace

File: enterprise/internal/insights/store/insight_store.go [hunk 8]
```diff
@@ -298,14 +329,63 @@ func (s *InsightStore) CreateView(ctx context.Context, view types.InsightView) (
...
+		return errors.New("unable to grant view permissions invalid insight view id")
+	} else if len(grants) == 0 {
+		return nil
+	}
+
+	values := make([]*sqlf.Query, 0, len(grants))
+	for _, grant := range grants {
+		values = append(values, grant.toQuery(view.ID))
+	}
+	q := sqlf.Sprintf(AddViewGrantsSql, sqlf.Join(values, ",\n"))
```
Comment (efritz) [inline enterprise/internal/insights/store/insight_store.go:355]: If grants can be *very* large (like over 32k / # cols) you could use the batch inserter utility in the internal/database package. Assuming the primary use of this method will be user-defined view grants, though, an not a mechanically generated list.

File: enterprise/internal/insights/store/insight_store.go [hunk 9]
```diff
@@ -298,14 +329,63 @@ func (s *InsightStore) CreateView(ctx context.Context, view types.InsightView) (
...
+	err := s.Exec(ctx, q)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+const AddViewGrantsSql = `
```
Comment (efritz) [inline enterprise/internal/insights/store/insight_store.go:363]: Do you want these queries exported?

File: internal/insights/insights.go [hunk 10]
```diff
@@ -209,11 +214,23 @@ func unmarshalIntegrated(raw json.RawMessage) (IntegratedInsights, error) {
...
-func (i IntegratedInsights) Insights() []SearchInsight {
+func (i IntegratedInsights) Insights(perms permissionAssociations) []SearchInsight {
 	results := make([]SearchInsight, 0)
 	for key, insight := range i {
 		insight.ID = key // the insight ID is the value of the dict key
+
+		// each setting is owned by either a user or an organization, which needs to be mapped when this insight is synced
+		// to preserve permissions semantics
+		insight.UserID = perms.userID
+		insight.OrgID = perms.orgID
```
Comment (efritz) [inline internal/insights/insights.go:232]: question: Does this mean that insights returned from graphql will always have _my_ user and _my_ org id?

**Review by efritz — APPROVED** [review]
🔒 👁️ 🚫

File: enterprise/internal/insights/store/insight_store.go (hunk 6 above)
Comment (coury-clark) [reply enterprise/internal/insights/store/insight_store.go:56]: Ah yeah - refactored this slightly and forgot to change the slice.

Comment (coury-clark) [inline enterprise/internal/insights/store/insight_store.go:56]:
```suggestion
	preds := make([]*sqlf.Query, 0, 3)
```

File: enterprise/internal/insights/store/insight_store.go (hunk 7 above)
Comment (coury-clark) [inline enterprise/internal/insights/store/insight_store.go:97]:
```suggestion
	return sqlf.Sprintf("(%s)", sqlf.Join(permsPreds, "OR"))
```

File: enterprise/internal/insights/store/insight_store.go (hunk 8 above)
Comment (coury-clark) [reply enterprise/internal/insights/store/insight_store.go:355]: In this case its all user operations, so I don't think there is any likely user interaction that would generate that many.

File: internal/insights/insights.go (hunk 10 above)
Comment (coury-clark) [reply internal/insights/insights.go:232]: It's pretty confusing but this is actually reading out of user / org settings (which is where insights were historically stored, before the backend became a thing). The GraphQL doesn't expose either of those things, this is just the sync from settings -> database.

File: enterprise/internal/insights/store/insight_store.go (hunk 9 above)
Comment (coury-clark) [reply enterprise/internal/insights/store/insight_store.go:363]: Nope - will fix
